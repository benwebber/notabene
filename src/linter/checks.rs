mod changes;
mod content;
pub(crate) mod preamble;
mod releases;
mod structure;
mod unreleased;

pub use changes::*;
pub use content::*;
pub use releases::*;
pub use structure::*;
pub use unreleased::*;
