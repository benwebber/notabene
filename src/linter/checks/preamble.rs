pub(crate) use crate::linter::Context;
pub(crate) use crate::linter::check::Check;
pub(crate) use crate::rule::Rule;
pub(crate) use crate::span::Span;

pub(crate) use crate::changelog::parsed;
