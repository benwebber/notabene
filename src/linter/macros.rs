macro_rules! invalid_span {
    ($check:ident) => {
        #[derive(Default)]
        pub struct $check;

        impl Check for $check {
            fn rule(&self) -> Rule {
                Rule::$check
            }

            fn visit_invalid_span(&mut self, context: &mut Context, span: &parsed::InvalidSpan) {
                if let parsed::InvalidSpan::$check(s) = span {
                    context.report(self.rule(), Some(*s));
                }
            }
        }
    };
}
