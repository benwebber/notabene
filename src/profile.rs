use std::collections::{HashMap, HashSet};

use crate::rule::Rule;

pub struct Profile {
    pub rules: HashSet<Rule>,
}

impl Profile {
    pub fn new(rules: HashSet<Rule>) -> Self {
        Self { rules }
    }

    pub fn is_enabled(&self, rule: Rule) -> bool {
        self.rules.contains(&rule)
    }
}

impl Default for Profile {
    fn default() -> Self {
        Self::from(Rule::ALL)
    }
}

impl<const N: usize> From<[Rule; N]> for Profile {
    fn from(rules: [Rule; N]) -> Self {
        Self::new(rules.iter().copied().collect())
    }
}
