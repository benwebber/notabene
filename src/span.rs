//! Work with spans within the source document.
use std::ops::{Deref, Range};

use serde::{Deserialize, Serialize};

mod iter;
mod locator;
mod ranged;
mod unist;

pub(crate) use iter::SpanIterator;
pub use locator::Locator;
pub(crate) use ranged::Ranged;
pub use unist::{Point, Position};

/// A span within the source document.
///
/// Use [`Locator`] to convert a `Span` to a [`Position`].
#[derive(
    Copy, Clone, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize,
)]
pub struct Span {
    /// The start offset of the span.
    pub start: usize,
    /// The end offset of the span.
    pub end: usize,
}

#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq, Serialize)]
pub(crate) struct Spanned<T> {
    pub span: Span,
    pub value: T,
}

impl Ranged<usize> for Span {
    fn range(&self) -> Range<usize> {
        (*self).into()
    }
}

impl<'a> Deref for Spanned<&'a str> {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        self.value
    }
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }

    /// Offset a span by `offset`.
    pub fn offset(&self, offset: usize) -> Self {
        Self {
            start: offset + self.start,
            end: offset + self.end,
        }
    }
}

impl From<Range<usize>> for Span {
    fn from(range: Range<usize>) -> Self {
        Self {
            start: range.start,
            end: range.end,
        }
    }
}

impl From<Span> for Range<usize> {
    fn from(span: Span) -> Self {
        span.start..span.end
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_offset() {
        let span = Span::new(0, 10);
        assert_eq!(span.offset(5), Span::new(5, 15));
    }

    #[test]
    fn test_range() {
        let span = Span::new(0, 10);
        assert_eq!(span.range(), span.into())
    }
}
