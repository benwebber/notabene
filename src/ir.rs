//! Intermediate representation (IR) of a changelog.
//!
//! Parsing a changelog returns an [`ir::Changelog`](crate::ir::Changelog).
//! This model differs from the [`Changelog`](crate::Changelog) in two significant ways:
//!
//!   * Elements include span information.
//!   * It allows multiple titles and unreleased sections.
//!
//! Preserving span information allows lint checks to include the context around diagnostics.
use crate::span::Span;

use serde::Serialize;

#[derive(Debug, Default, Eq, Hash, PartialEq, Serialize)]
pub(crate) struct Spanned<T> {
    pub span: Span,
    pub value: T,
}

#[derive(Debug, Default, Serialize)]
pub(crate) struct Changelog<'a> {
    pub sections: Vec<Section<'a>>,
    pub broken_links: Vec<Span>,
}

#[derive(Debug, Serialize)]
pub enum Section<'a> {
    Title(Spanned<&'a str>),
    Unreleased(Unreleased<'a>),
    Release(Release<'a>),
    Invalid(InvalidSection),
    // TODO: If Title held something like Vec<Inline>, it would be possible
    // to get rid of InvalidTitle and validate Title in the linter.
    InvalidTitle(InvalidTitle),
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct Unreleased<'a> {
    pub heading_span: Span,
    // TODO: Get span.
    pub url: Option<String>,
    pub changes: Vec<Changes<'a>>,
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct Release<'a> {
    pub heading_span: Span,
    pub version: Spanned<&'a str>,
    // TODO: Get span.
    pub url: Option<String>,
    pub date: Option<Spanned<&'a str>>,
    pub yanked: Option<Spanned<&'a str>>,
    pub changes: Vec<Changes<'a>>,
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct Changes<'a> {
    pub heading_span: Span,
    pub kind: Spanned<&'a str>,
    pub changes: Vec<Spanned<&'a str>>,
}

impl<T> Spanned<T> {
    pub(crate) fn new(span: Span, value: T) -> Self {
        Self { value, span }
    }

    pub(crate) fn into_inner(self) -> T {
        self.value
    }
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct InvalidSection {
    pub heading_span: Span,
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct InvalidTitle {
    pub heading_span: Span,
}
