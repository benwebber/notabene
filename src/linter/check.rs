use crate::diagnostic::Diagnostic;
use crate::ir;
use crate::location::Location;
use crate::rule::Rule;
use crate::span::Span;

pub(crate) trait Check {
    /// Return the `Rule` this check evaluates.
    fn rule(&self) -> Rule;

    /// Return spans that this check matched.
    fn spans(&self) -> &[Span] {
        &[]
    }

    /// Evaluate this check for the entire `Changelog`.
    fn visit_changelog(&mut self, _changelog: &ir::Changelog) {}

    /// Evaluate this check for a `Section`.
    fn visit_section(&mut self, _section: &ir::Section) {}

    /// Evaluate this check for a `Changes`.
    fn visit_changes(&mut self, _changes: &ir::Changes) {}

    /// Publish diagnostics generated by this check.
    fn diagnostics(&self) -> Vec<Diagnostic> {
        self.spans()
            .iter()
            .map(|span| Diagnostic::new(self.rule(), Some(Location::Span(*span))))
            .collect()
    }
}
