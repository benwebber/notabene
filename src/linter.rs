//! Linter implementation.
use crate::diagnostic::Diagnostic;
use crate::ir::{Changelog, Changes, Section};
use crate::profile::Profile;
use crate::rule::Rule;
use crate::span::Span;

mod check;
mod checks;

use check::Check;

pub fn lint(changelog: &Changelog, profile: &Profile) -> Vec<Diagnostic> {
    let mut diagnostics = Vec::new();

    let mut checks: Vec<_> = checks()
        .into_iter()
        .filter(|check| profile.is_enabled(check.rule()))
        .collect();

    for section in &changelog.sections {
        for check in checks.iter_mut() {
            check.visit_section(section);
            match section {
                Section::Unreleased(unreleased) => {
                    for changes in &unreleased.changes {
                        check.visit_changes(changes);
                    }
                }
                Section::Release(release) => {
                    for changes in &release.changes {
                        check.visit_changes(changes);
                    }
                }
                _ => {}
            }
        }
    }

    for check in checks.iter_mut() {
        diagnostics.append(&mut check.diagnostics());
    }

    diagnostics
}

fn checks() -> Vec<Box<dyn Check>> {
    vec![
        // E000 Document
        Box::new(checks::MissingTitle::default()),
        Box::new(checks::DuplicateTitle::default()),
        Box::new(checks::UnreleasedNotFirst::default()),
        //// E100 Unreleased
        Box::new(checks::MissingUnreleased::default()),
        Box::new(checks::DuplicateUnreleased::default()),
        //// E200 Release
        Box::new(checks::InvalidDate::default()),
        Box::new(checks::InvalidYanked::default()),
        Box::new(checks::MissingDate::default()),
        Box::new(checks::ReleaseOutOfOrder::default()),
        Box::new(checks::DuplicateVersion::default()),
        //// E300 Changes
        Box::new(checks::InvalidChangeType::default()),
        Box::new(checks::DuplicateChangeType::default()),
        //// E400 Content
        Box::new(checks::EmptySection::default()),
    ]
}
