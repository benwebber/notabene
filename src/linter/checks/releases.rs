use time::Date;
use time::macros::format_description;

use super::preamble::*;

#[derive(Default)]
pub struct InvalidDate {
    spans: Vec<Span>,
}

impl Check for InvalidDate {
    fn rule(&self) -> Rule {
        Rule::InvalidDate
    }

    fn spans(&self) -> &[Span] {
        self.spans.as_slice()
    }

    fn visit_section(&mut self, section: &Section) {
        let format = format_description!("[year]-[month]-[day]");
        if let Section::Release(release) = section {
            if let Some(spanned) = &release.date {
                if Date::parse(&spanned.value, &format).is_err() {
                    self.spans.push(spanned.span);
                }
            }
        }
    }
}

#[derive(Default)]
pub struct InvalidYanked {
    spans: Vec<Span>,
}

impl Check for InvalidYanked {
    fn rule(&self) -> Rule {
        Rule::InvalidYanked
    }

    fn spans(&self) -> &[Span] {
        self.spans.as_slice()
    }

    fn visit_section(&mut self, section: &Section) {
        if let Section::Release(release) = section {
            if let Some(spanned) = &release.yanked {
                if spanned.value != "[YANKED]" {
                    self.spans.push(spanned.span);
                }
            }
        }
    }
}

#[derive(Default)]
pub struct MissingDate {
    spans: Vec<Span>,
}

impl Check for MissingDate {
    fn rule(&self) -> Rule {
        Rule::MissingDate
    }

    fn spans(&self) -> &[Span] {
        self.spans.as_slice()
    }

    fn visit_section(&mut self, section: &Section) {
        if let Section::Release(release) = section {
            if release.date.is_none() {
                self.spans.push(release.heading_span);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_yaml_snapshot;

    use crate::ir::*;
    use crate::linter::lint;
    use crate::profile::Profile;
    use crate::span::Span;

    #[test]
    fn test_invalid_date() {
        let profile = Profile::new(&[Rule::InvalidDate]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![
                Section::Release(Release {
                    ..Default::default()
                }),
                Section::Release(Release {
                    date: Some(Spanned::new(Span::new(0, 9), "2038-01-19".to_string())),
                    ..Default::default()
                }),
                Section::Release(Release {
                    date: Some(Spanned::new(Span::new(1, 10), "2001-01-00".to_string())),
                    ..Default::default()
                }),
                Section::Release(Release {
                    date: Some(Spanned::new(Span::new(2, 5), "foo".to_string())),
                    ..Default::default()
                }),
            ],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }

    #[test]
    fn test_invalid_yanked() {
        let profile = Profile::new(&[Rule::InvalidYanked]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![
                Section::Release(Release {
                    ..Default::default()
                }),
                Section::Release(Release {
                    yanked: Some(Spanned::new(Span::new(0, 9), "[YANKED]".to_string())),
                    ..Default::default()
                }),
                Section::Release(Release {
                    yanked: Some(Spanned::new(Span::new(1, 10), "[ZANKED]".to_string())),
                    ..Default::default()
                }),
            ],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }

    #[test]
    fn test_missing_date() {
        let profile = Profile::new(&[Rule::MissingDate]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![
                Section::Release(Release {
                    date: Some(Spanned::new(Span::new(0, 11), "2025-01-01".to_string())),
                    ..Default::default()
                }),
                Section::Release(Release {
                    heading_span: Span::new(1, usize::MAX),
                    ..Default::default()
                }),
            ],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }
}
