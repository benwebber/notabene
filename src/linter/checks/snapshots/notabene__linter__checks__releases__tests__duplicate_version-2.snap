---
source: src/linter/checks/releases.rs
expression: "lint(&changelog, &profile)"
---
- rule: DuplicateVersion
  span:
    start: 1
    end: 18446744073709551615
  path: ~
